packages/yew/src/html/component/lifecycle.rs

```diff
@@ -1,22 +1,97 @@
 //! Component lifecycle module
 
 use super::{AnyScope, BaseComponent, Scope};
-use crate::html::RenderError;
+use crate::html::{RenderError, RenderResult};
 use crate::scheduler::{self, Runnable, Shared};
 use crate::suspense::{Suspense, Suspension};
 use crate::virtual_dom::{VDiff, VNode};
 use crate::Callback;
 use crate::{Context, NodeRef};
 #[cfg(feature = "ssr")]
 use futures::channel::oneshot;
+use std::any::Any;
 use std::rc::Rc;
 use web_sys::Element;
 
-pub(crate) struct ComponentState<COMP: BaseComponent> {
-    pub(crate) component: Box<COMP>,
-    pub(crate) root_node: VNode,
+pub(crate) struct CompStateInner<COMP>
+where
+    COMP: BaseComponent,
+{
+    pub(crate) component: COMP,
+    pub(crate) context: Context<COMP>,
+}
+
+pub(crate) trait Stateful {
+    fn view(&self) -> RenderResult<VNode>;
+    fn rendered(&mut self, first_render: bool);
+    fn destroy(&mut self);
+
+    fn any_scope(&self) -> AnyScope;
+
+    fn flush_messages(&mut self) -> bool;
+    fn props_changed(&mut self, props: Rc<dyn Any>) -> bool;
+
+    fn as_any(&self) -> &dyn Any;
+    fn as_any_mut(&mut self) -> &mut dyn Any;
+}
+
+impl<COMP> Stateful for CompStateInner<COMP>
+where
+    COMP: BaseComponent,
+{
+    fn view(&self) -> RenderResult<VNode> {
+        self.component.view(&self.context)
+    }
+
+    fn rendered(&mut self, first_render: bool) {
+        self.component.rendered(&self.context, first_render)
+    }
+
+    fn destroy(&mut self) {
+        self.component.destroy(&self.context);
+    }
+
+    fn any_scope(&self) -> AnyScope {
+        self.context.link().clone().into()
+    }
+
+    fn flush_messages(&mut self) -> bool {
+        self.context
+            .link()
+            .pending_messages
+            .drain()
+            .into_iter()
+            .fold(false, |acc, msg| {
+                self.component.update(&self.context, msg) || acc
+            })
+    }
+
+    fn props_changed(&mut self, props: Rc<dyn Any>) -> bool {
+        let props = match Rc::downcast::<COMP::Properties>(props) {
+            Ok(m) => m,
+            _ => return false,
+        };
 
-    context: Context<COMP>,
+        if self.context.props != props {
+            self.context.props = Rc::clone(&props);
+            self.component.changed(&self.context)
+        } else {
+            false
+        }
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+    fn as_any_mut(&mut self) -> &mut dyn Any {
+        self
+    }
+}
+
+pub(crate) struct ComponentState {
```

Comment:
Removing the generic arguments from `ComponentState` is the main reason for reducing the bundle size, isn't it? Not saying that is bad, I just want to understand where the reduction is coming from.

Reply:
Like stated in the description, it's from the `*Runner`'s implementation below the component state. If the runners contain a generic, then the entire implementation needs to be duplicated for every component.